src/Chessboard.cpp

```diff
@@ -0,0 +1,313 @@
+        return 13; // can't capture own piece
+    }
+
+    if (!startPiece->isValidMove(endRow, endCol)) {
+        return 21; // invalid move
+    }
+
+    Chessboard copy = *this;
+    copy.board_[endRow][endCol] = startPiece;
+    copy.board_[startRow][startCol] = nullptr;
+    startPiece->setRow(endRow);
+    startPiece->setCol(endCol);
+
+    if (copy.isCheck(currentPlayer_)) {
+        return 31; // moving will cause checkmate
+    }
+
+    board_[endRow][endCol] = startPiece;
+    board_[startRow][startCol] = nullptr;
+    startPiece->setRow(endRow);
+    startPiece->setCol(endCol);
+    if (isCheck(currentPlayer_)) {
+        return 31; // moving will cause checkmate
+    }
+
+    board_[endRow][endCol] = startPiece;
+    board_[startRow][startCol] = nullptr;
+    startPiece->setRow(endRow);
+    startPiece->setCol(endCol);
+
+    if (isCheck(currentPlayer_)) {
+        return 41; // legal move and caused check
+    }
+    else {
+        currentPlayer_ = (currentPlayer_ == PieceColor::White) ? PieceColor::Black : PieceColor::White;
+        return 42; // legal move and next turn
+    }
+}
+
+    
+bool Chessboard::isCheck(PieceColor color) {
+// find the king
+    int kingRow = -1, kingCol = -1;
+    for (int i = 0; i < 8; i++) {
+        for (int j = 0; j < 8; j++) {
+            Piece* piece = board_[i][j];
+            if (piece != nullptr && piece->getType() == PieceType::King && piece->getColor() == color) {
+                kingRow = i;
+                kingCol = j;
+                break;
+            }
+        }
+        if (kingRow != -1 && kingCol != -1) {
+            break;
+        }
+    }
+
+    // check for rooks and queens attacking from the left or right
+    for (int i = kingCol + 1; i < 8; i++) {
+        Piece* piece = board_[kingRow][i];
+        if (piece != nullptr) {
+            if (piece->getColor() == color) {
+                break;
+            }
+            else if (piece->getType() == PieceType::Rook || piece->getType() == PieceType::Queen) {
+                return true;
+            }
+            else {
+                break;
+            }
+        }
+    }
+    for (int i = kingCol - 1; i >= 0; i--) {
+        Piece* piece = board_[kingRow][i];
+        if (piece != nullptr) {
+            if (piece->getColor() == color) {
+                break;
+            }
+            else if (piece->getType() == PieceType::Rook || piece->getType() == PieceType::Queen) {
+                return true;
+            }
+            else {
+                break;
+            }
+        }
+    }
+
+    // check for rooks and queens attacking from above or below
+    for (int i = kingRow + 1; i < 8; i++) {
+        Piece* piece = board_[i][kingCol];
+        if (piece != nullptr) {
+            if (piece->getColor() == color) {
+                break;
+            }
+            else if (piece->getType() == PieceType::Rook || piece->getType() == PieceType::Queen) {

```

Comment:
The logic for checking if a piece is a rook, bishop, queen, or knight is repeated multiple times. You can encapsulate this logic into a separate function to make the code more readable and maintainable.